File: lib/quantbot/core/portfolio.py
"""Portfolio state, position, and order types."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import UTC, datetime
from enum import Enum
from typing import Any


class OrderSide(Enum):
    BUY = "BUY"
    SELL = "SELL"


@dataclass
class Position:
    instrument: str
    quantity: float  # positive = long, negative = short
    avg_entry_price: float
    point_value: float = 1.0

    @property
    def notional(self) -> float:
        return abs(self.quantity) * self.avg_entry_price * self.point_value

    def unrealised_pnl(self, current_price: float) -> float:
        return self.quantity * (current_price - self.avg_entry_price) * self.point_value


@dataclass
class Order:
    instrument: str
    side: OrderSide
    quantity: float  # always positive
    timestamp: datetime = field(default_factory=lambda: datetime.now(UTC))
    limit_price: float | None = None
    metadata: dict[str, Any] = field(default_factory=dict)


@dataclass
class Fill:
    order: Order
    fill_price: float
    fill_quantity: float
    timestamp: datetime = field(default_factory=lambda: datetime.now(UTC))
    slippage_bps: float = 0.0


@dataclass
class PortfolioState:
    """Snapshot of portfolio at a point in time."""

    cash: float
    positions: dict[str, Position] = field(default_factory=dict)
    timestamp: datetime = field(default_factory=lambda: datetime.now(UTC))

    @property
    def nav(self) -> float:
        """Net asset value (requires current prices passed separately for real use).
        For backtest snapshots, positions carry mark prices in avg_entry_price after re-mark."""
        pos_value = sum(
            p.quantity * p.avg_entry_price * p.point_value
            for p in self.positions.values()
        )
        return self.cash + pos_value

    def gross_exposure(self, prices: dict[str, float] | None = None) -> float:
        total = 0.0
        for sym, pos in self.positions.items():
            px = prices.get(sym, pos.avg_entry_price) if prices else pos.avg_entry_price
            total += abs(pos.quantity) * px * pos.point_value
        return total

    def net_exposure(self, prices: dict[str, float] | None = None) -> float:
        total = 0.0
        for sym, pos in self.positions.items():
            px = prices.get(sym, pos.avg_entry_price) if prices else pos.avg_entry_price
            total += pos.quantity * px * pos.point_value
        return total

File: src/core/portfolio.rs
use std::collections::HashMap;

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum OrderSide {
    Buy,
    Sell,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Position {
    pub instrument: String,
    pub quantity: f64,
    pub avg_entry_price: f64,
    pub point_value: f64,
}

impl Position {
    pub fn new(instrument: String, quantity: f64, avg_entry_price: f64) -> Self {
        Self {
            instrument,
            quantity,
            avg_entry_price,
            point_value: 1.0,
        }
    }

    pub fn notional(&self) -> f64 {
        self.quantity.abs() * self.avg_entry_price * self.point_value
    }

    pub fn unrealised_pnl(&self, current_price: f64) -> f64 {
        self.quantity * (current_price - self.avg_entry_price) * self.point_value
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Order {
    pub instrument: String,
    pub side: OrderSide,
    pub quantity: f64,
    pub timestamp: DateTime<Utc>,
    pub limit_price: Option<f64>,
}

impl Order {
    pub fn new(instrument: String, side: OrderSide, quantity: f64) -> Self {
        Self {
            instrument,
            side,
            quantity,
            timestamp: Utc::now(),
            limit_price: None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Fill {
    pub order: Order,
    pub fill_price: f64,
    pub fill_quantity: f64,
    pub timestamp: DateTime<Utc>,
    pub slippage_bps: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PortfolioState {
    pub cash: f64,
    pub positions: HashMap<String, Position>,
    pub timestamp: DateTime<Utc>,
}

impl PortfolioState {
    pub fn new(cash: f64) -> Self {
        Self {
            cash,
            positions: HashMap::new(),
            timestamp: Utc::now(),
        }
    }

    pub fn nav(&self) -> f64 {
        let pos_value: f64 = self
            .positions
            .values()
            .map(|p| p.quantity * p.avg_entry_price * p.point_value)
            .sum();
        self.cash + pos_value
    }

    pub fn gross_exposure(&self, prices: Option<&HashMap<String, f64>>) -> f64 {
        self.positions
            .iter()
            .map(|(sym, pos)| {
                let px = prices
                    .and_then(|p| p.get(sym))
                    .copied()
                    .unwrap_or(pos.avg_entry_price);
                pos.quantity.abs() * px * pos.point_value
            })
            .sum()
    }

    pub fn net_exposure(&self, prices: Option<&HashMap<String, f64>>) -> f64 {
        self.positions
            .iter()
            .map(|(sym, pos)| {
                let px = prices
                    .and_then(|p| p.get(sym))
                    .copied()
                    .unwrap_or(pos.avg_entry_price);
                pos.quantity * px * pos.point_value
            })
            .sum()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn position_notional_and_pnl() {
        let pos = Position {
            instrument: "GC=F".into(),
            quantity: 2.0,
            avg_entry_price: 2000.0,
            point_value: 100.0,
        };
        assert!((pos.notional() - 400_000.0).abs() < 1e-10);
        assert!((pos.unrealised_pnl(2050.0) - 10_000.0).abs() < 1e-10);
    }

    #[test]
    fn short_position_pnl() {
        let pos = Position {
            instrument: "SPY".into(),
            quantity: -10.0,
            avg_entry_price: 500.0,
            point_value: 1.0,
        };
        // Price dropped to 490 → profit for short
        assert!((pos.unrealised_pnl(490.0) - 100.0).abs() < 1e-10);
    }

    #[test]
    fn portfolio_nav_and_exposure() {
        let mut state = PortfolioState::new(100_000.0);
        state
            .positions
            .insert("SPY".into(), Position::new("SPY".into(), 10.0, 500.0));
        state.positions.insert(
            "GLD".into(),
            Position {
                instrument: "GLD".into(),
                quantity: -5.0,
                avg_entry_price: 200.0,
                point_value: 1.0,
            },
        );

        // NAV = 100_000 + (10*500) + (-5*200) = 100_000 + 5000 - 1000 = 104_000
        assert!((state.nav() - 104_000.0).abs() < 1e-10);

        // Gross exposure without prices: |10|*500 + |-5|*200 = 5000 + 1000 = 6000
        assert!((state.gross_exposure(None) - 6_000.0).abs() < 1e-10);

        // Net exposure without prices: 10*500 + (-5)*200 = 5000 - 1000 = 4000
        assert!((state.net_exposure(None) - 4_000.0).abs() < 1e-10);
    }
}
